File: Problem171_divide-two-integers.java
// Time Complexity: O(log n)
// Space Complexity: O(1)
class Solution {
    public int divide(int dividend, int divisor) {
        if(divisor == 0 && dividend != 0) return Integer.MAX_VALUE;
        if(dividend == Integer.MIN_VALUE && divisor == -1) return Integer.MAX_VALUE;
        // 25
        // 4 >> 8 >> 16 >> 32
        int result = 0;
        long ldividend = Math.abs((long)dividend);
        long ldivisor = Math.abs((long)divisor);
        while(ldividend >= ldivisor) { // 9 4 >> 8
            int shifts = 0;
            while(ldividend >= (ldivisor << shifts)) { // 9 4 >> 8 >> 16
                shifts++;
            }
            shifts--;
            result += 1 << shifts; // 4 + 2 = 6
            ldividend = ldividend - (ldivisor << shifts); // 25 - 16 = 9 , 9 - 8 = 1
        }
        if(dividend > 0 && divisor > 0) return result;
        if(dividend < 0 && divisor < 0) return result;
        return -result;
    }
}

File: Problem-171 Divide Two Integers.py
# 29. Divide Two Integers
# https://leetcode.com/problems/divide-two-integers/

# Time complexity: O(log n)
# Space Complexity: O(1)

class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        intMax = (2**31) - 1
        intMin = -2**31
        
        if dividend == intMin and divisor == -1:
            return intMax
        
        newDividend = abs(dividend)
        newDivisor = abs(divisor)
        
        result = 0
        
        while newDividend >= newDivisor:
            shifts = 0
            temp = newDivisor
            
            while temp <= newDividend:
                temp = temp << 1
                shifts += 1
            
            temp = temp // 2
            shifts -= 1
            
            result += 2**shifts
            
            newDividend -= temp
        
        if dividend < 0 and divisor > 0 or dividend > 0 and divisor < 0: 
            return -result
        return result